core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -955,6 +940,80 @@ public void run() {
       return pendingCall;
     }
 
+    // Must run in SynchronizationContext.
+    private void drainPendingCalls() {
+      if (pendingCalls == null) {
+        return;
+      }
+      for (RealChannel.PendingCall<?, ?> pendingCall : pendingCalls) {
+        pendingCall.reprocess();
+      }
+    }
+
+    void shutdown() {
+      if (!shutdown.compareAndSet(false, true)) {
+        return;
+      }
+      // Put gotoState(SHUTDOWN) as early into the syncContext's queue as possible.
+      // delayedTransport.shutdown() may also add some tasks into the queue. But some things inside
+      // delayedTransport.shutdown() like setting delayedTransport.shutdown = true are not run in
+      // the syncContext's queue and should not be blocked, so we do not drain() immediately here.
+      syncContext.executeLater(new Runnable() {
+        @Override
+        public void run() {
+          channelLogger.log(ChannelLogLevel.INFO, "Entering SHUTDOWN state");
+          channelStateManager.gotoState(SHUTDOWN);
+        }
+      });
+      if (configSelector.get() != INITIAL_PENDING_SELECTOR && pendingCalls == null) {
+        uncommittedRetriableStreamsRegistry.onShutdown(SHUTDOWN_STATUS);
+        syncContext.drain();
+        return;
+      }
+
+      final class Shutdown implements Runnable {
+        @Override
+        public void run() {
+          if (pendingCalls == null) {
+            if (configSelector.get() == INITIAL_PENDING_SELECTOR) {
+              configSelector.set(null);
+            }
+            uncommittedRetriableStreamsRegistry.onShutdown(SHUTDOWN_STATUS);
+          }
+        }
+      }
+
+      syncContext.execute(new Shutdown());
+    }
+
+    void shutdownNow() {
+      if (configSelector.get() != INITIAL_PENDING_SELECTOR && pendingCalls == null) {
+        uncommittedRetriableStreamsRegistry.onShutdownNow(SHUTDOWN_NOW_STATUS);
+      }
+      final class ShutdownNow implements Runnable {
+        @Override
+        public void run() {
+          if (shutdownNowed) {
+            return;
+          }
+          shutdownNowed = true;
+          if (configSelector.get() == INITIAL_PENDING_SELECTOR) {
+            configSelector.set(null);
+          }
+          if (pendingCalls == null) {
+            uncommittedRetriableStreamsRegistry.onShutdownNow(SHUTDOWN_NOW_STATUS);
```

Comment:
Why isn't this called unconditionally?

Reply:
Oh, I was planning to do it unconditionally in [simplify PendingCallRemoval](https://github.com/grpc/grpc-java/pull/7362/commits/879dd06285349cc121e16a426401b3c7931b25a6) but forgot.